src/lib/tools.ts

```diff
@@ -0,0 +1,245 @@
+import * as Bluebird from 'bluebird';
+import * as _ from 'lodash';
+import * as fs from 'mz/fs';
+import { dirname, isAbsolute, join, resolve } from 'path';
+import * as tunnel from 'tunnel-ssh';
+import {
+	FileLoadError,
+	NotImplementedError,
+	UnsupportedError,
+	URILoadError,
+} from './error-types';
+
+import { TimeoutError } from 'bluebird';
+import * as yamljs from 'yamljs';
+import { ConfigMap } from './controllers/config-store';
+
+export const tunnelAsync = Bluebird.promisify(tunnel);
+
+export async function getDirectories(path: string): Promise<string[]> {
+	const directories = (await fs.readdir(path)) as string[];
+	return directories.filter(async (name: string) => {
+		const itemPath = join(path, name);
+		const itemStat = await fs.stat(itemPath);
+		return itemStat.isDirectory();
+	});
+}
+
+/**
+ * Returns an absolute path for a path, when in basePath
+ * @param {string} path: The file path, (or an absolute path)
+ * @param {string} basePath: The base path
+ * @returns {string} An asolute path
+ */
+export function getAbsolutePath(path: string, basePath: string): string {
+	return isAbsolute(path) ? path : join(basePath, path);
+}
+
+export function getAbsoluteURI(uri: string, basePath: string): string {
+	if (gitURI(uri)) {
+		throw new NotImplementedError('Git URI support not implemented yet');
+	} else if (localPathURI(uri)) {
+		return getAbsolutePath(uri, basePath);
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+export async function loadFromFile(
+	filePath: string,
+	errorMessage: string = '',
+): Promise<any> {
+	try {
+		const buffer = await fs.readFile(filePath);
+		return yamljs.parse(buffer.toString('utf8'));
+	} catch (e) {
+		throw new FileLoadError(errorMessage + e.message);
+	}
+}
+
+export function gitURI(uri: string): boolean {
+	return /((git|ssh|http|https)|(git@[\w\.]+))(:(\/\/)?)([\w\.@\:/\-]+)/.test(
+		uri,
+	);
+}
+
+export function localPathURI(uri: string): boolean {
+	return /^([a-zA-Z0-9_/\-.])+$/.test(uri);
+}
+
+export async function loadFromURI(
+	URI: string,
+	path: string,
+	errorMessage?: string,
+): Promise<any> {
+	// TODO: support git URI
+	if (gitURI(URI)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	}
+	try {
+		if (localPathURI(URI)) {
+			return await loadFromFile(join(URI, path), errorMessage);
+		}
+	} catch (e) {
+		throw new URILoadError(`Error loading ${path} from ${URI}\n${e.message}`);
+	}
+}
+
+export function getBasePath(path: string): string {
+	return dirname(resolve(path));
+}
+
+export function convertRelativePaths({
+	conf,
+	basePath,
+}: {
+	conf: any;
```

Comment:
At the very least this is an interface with string keys which consist of objects with `string:value`?